src/tla/modules/raft.tla

```diff
@@ -0,0 +1,394 @@
+\*               /\ leader[i] # entry.replica_id
+\*               /\ leader[i] = Nil
+\*               /\ RaftFollowLeader(i, entry.replica_id)
+\*
+RaftProcessMsg(i, entry) ==
+    IF entry.body.term >= volatileTerm[i]
+    THEN LET raftVarsInit == RaftVars(i)
+             raftVarsTerm ==
+                IF entry.body.term > raftVarsInit.volatileTerm[i]
+                THEN RaftScheduleNewTermVars(raftVars, entry.body.term)
+                ELSE raftVarsInit
+             raftVarsSeen == [raftVarsTerm EXCEPT
+                    !.leaderWitnessMap[entry.replica_id] =
+                        entry.body.is_leader_seen]
+             raftVarsVote ==
+                IF entry.body.vote # 0
+                THEN IF raftVarsSeen.state \in {Follower, Leader}
+                     THEN IF /\ raftVarsSeen.leader = Nil
+                             /\ entry.body.vote # i
+                             /\ entry.body.state = Candidate
+                             /\ raftVarsSeen.volatileVote = Nil
+                             /\ RaftCanVoteFor(i, entry.body.vclock)
+                          THEN RaftScheduleNewVoteVars(raftVarsSeen,
+                                entry.replica_id, entry.body.vclock)
+                          ELSE raftVarsSeen
+                     ELSE \* state = Candidate
+                          IF entry.body.vote = i
+                          THEN LET raftVarsTmp == [raftVarsSeen EXCEPT
+                            !.votesReceived = @ \cup {entry.replica_id}]
+                               IN IF Cardinality(raftVarsTmp.votesReceived) >=
+                                    ElectionQuorum
+                                  THEN RaftBecomeLeaderVars(raftVarsTmp, i)
+                                  ELSE raftVarsTmp
+                          ELSE raftVarsSeen
+                ELSE raftVarsSeen
+             raftVarsFinal == IF entry.body.state # Leader
+                              THEN IF raftVarsVote.leader = entry.replica_id
+                                   THEN [raftVarsVote EXCEPT
+                                            !.leader = Nil,
+                                            !.leaderWitnessMap[i] = FALSE,
+                                            !.isBroadcastScheduled = TRUE]
+                                   ELSE raftVarsVote
+                              ELSE \* entry.body.state = Leader
+                                   IF raftVarsVote.leader # entry.replica_id
+                                   THEN \* raft_sm_follow_leader
+                                        [raftVarsVote EXCEPT
+                                            !.state = Follower,
+                                            !.leader = entry.replica_id,
+                                            !.isBroadcastScheduled = TRUE]
+                                   ELSE raftVarsVote
+         IN RaftApplyVars(i, raftVarsFinal)
+    ELSE UNCHANGED <<raftVars>>
+
+RaftNext(servers) ==
```

Comment:
Original spec made by Diego Ongaro [^1] simulates unreliable network by duplicate and dropped messages. Will you do the same?

[^1]: https://github.com/ongardie/raft.tla/blob/master/raft.tla

Reply:
I don't think we can, tarantool strongly relies on the order of messages, if it'll get lsn 100 after 1 that is an error. We may consider mixing raft msgs, but this will just cause split brain

Reply:
Spec should check that a system designed and behaves correctly. I will not insist, but checking behavior with unreliable network would be useful.
Just close a comment if you still do not agree and don't want to implement it.

Reply:
https://github.com/tarantool/tarantool/issues/11385